core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -157,18 +157,57 @@ __global__ static void cuda_parallel_launch_constant_or_global_memory(
   driver();
 }
 
-template <class DriverType, unsigned int maxTperB, unsigned int minBperSM>
-__global__
-__launch_bounds__(maxTperB, minBperSM) static void cuda_parallel_launch_constant_or_global_memory(
-    const DriverType* driver_ptr) {
-  const DriverType& driver =
-      driver_ptr != nullptr
-          ? *driver_ptr
-          : *((const DriverType*)kokkos_impl_cuda_constant_memory_buffer);
+//==============================================================================
+// <editor-fold desc="Some helper functions for launch code readability"> {{{1
 
-  driver();
+inline bool is_empty_launch(dim3 const& grid, dim3 const& block) {
+  return (grid.x == 0) || ((block.x * block.y * block.z) == 0);
+}
+
+inline void check_shmem_request(CudaInternal const* cuda_instance, int shmem) {
+  if (cuda_instance->m_maxShmemPerBlock < shmem) {
+    Kokkos::Impl::throw_runtime_exception(
+        std::string("CudaParallelLaunch (graph node) FAILED: shared memory"
+                    "request is too large"));
+  }
+}
+
+template <class KernelFuncPtr>
+inline void configure_shmem_preference(KernelFuncPtr const& func,
+                                       bool prefer_shmem) {
+#ifndef KOKKOS_ARCH_KEPLER
+  // On Kepler the L1 has no benefit since it doesn't cache reads
+  static bool cache_config_set = false;
+  static bool cache_config_preference_cached;
+  if (!cache_config_set) {
+    cache_config_preference_cached = prefer_shmem;
+    CUDA_SAFE_CALL(cudaFuncSetCacheConfig(
+        func,
+        (prefer_shmem ? cudaFuncCachePreferShared : cudaFuncCachePreferL1)));
+    cache_config_set = true;
+  } else {
+    KOKKOS_ASSERT(cache_config_preference_cached == prefer_shmem);
+    // use the variable in case we're not compiling with contracts
+    if (cache_config_preference_cached) {
+    };
```

Comment:
Why did you chose this over the usual cast to `void`?

Reply:
I think the compiler complained for some reason when I tried the usual thing. I can try going back and see what happens, but that's what I remember.

Reply:
resolved by ed54a98529d0f58adb83c9a4b5158acfd504f38c